### PR TITLE
Add a Runtime.IsARM64CallingConvention property.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Determining whether we should use the ARM64 calling convention in P/Invokes
gets more complicated with ARM64_32 (which for our purposes is a 32-bit
architecture).

So add a property on the Runtime class to avoid code duplication, and teach
the linker to optimize any calls to this property to a constant value whenever
possible (and the method is marked as optimizable).

Also change our code to use this new property, and make the corresponding
methods optimizable.

Some shuffling in mmp was required, which meant a little bit more code is now
shared between mtouch and mmp.

### CODE REVIEW

File: src/ObjCRuntime/CpuArchitecture.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+namespace ObjCRuntime
+{
```
Comment (VincentDondain) [inline src/ObjCRuntime/CpuArchitecture.cs]: Minor: but wrong coding guideline (should be same line). That's for this entire PR so it sounds like your settings are incorrect (:

Comment (VincentDondain) [reply src/ObjCRuntime/CpuArchitecture.cs]: Clarification: for at least your `enum` and `struct`, those should be same line.

File: tools/common/Optimizations.cs [hunk 2]
```diff
@@ -249,6 +256,9 @@ public void Initialize (Application app)
 				RemoveUnsupportedILForBitcode = app.Platform == Utils.ApplePlatform.WatchOS && app.IsDeviceBuild;
 			}
 #endif
+			// Always enable Runtime.CpuArchitecture inlining.
```
Comment (spouliot) [inline tools/common/Optimizations.cs]: That's more enabling _by default_ than _always_ enabled :)

File: tools/linker/CoreOptimizeGeneratedCode.cs [hunk 3]
```diff
@@ -44,9 +45,34 @@ public abstract class CoreOptimizeGeneratedCode : ExceptionalSubStep {
...
 
+		bool? is_arm64_calling_convention;
+
 		public CoreOptimizeGeneratedCode (LinkerOptions options)
 		{
 			Options = options;
+
+			if (Optimizations.InlineIsARM64CallingConvention == true) { 
+				if (options.Target.Abis.Count == 1) {
+					// We can only inline Runtime.InlineIsARM64CallingConvention if the generated code will execute on a single architecture (not if building for armv7+armv7s for instance).
```
Comment (spouliot) [inline tools/linker/CoreOptimizeGeneratedCode.cs]: why ? is `armv7+armv7s` then we know arm64 calling conversion is always `false` right ?

File: tests/linker/mac/link all/link all-mac.csproj [hunk 4]
```diff
@@ -81,6 +81,9 @@
     <Compile Include="..\..\CommonLinkAllTest.cs">
       <Link>CommonLinkAllTest.cs</Link>
     </Compile>
+    <Compile Include="..\..\ios\link all\ILReader.cs">
+      <Link>ILReader.cs</Link>
+    </Compile>
```
Comment (spouliot) [inline tests/linker/mac/link all/link all-mac.csproj]: should be moved to `common` if used for the Mac

File: src/ObjCRuntime/CpuArchitecture.cs (hunk 1 above)
Comment (rolfbjarne) [reply src/ObjCRuntime/CpuArchitecture.cs]: Fixed.

File: tools/common/Optimizations.cs (hunk 2 above)
Comment (rolfbjarne) [reply tools/common/Optimizations.cs]: I've updated the comment.

File: tools/linker/CoreOptimizeGeneratedCode.cs (hunk 3 above)
Comment (rolfbjarne) [reply tools/linker/CoreOptimizeGeneratedCode.cs]: You're correct, I've changed the logic accordingly.

File: tests/linker/mac/link all/link all-mac.csproj (hunk 4 above)
Comment (rolfbjarne) [reply tests/linker/mac/link all/link all-mac.csproj]: Done

**Pull request merged**